go/src/koding/kites/kloud/stack/provider/builder.go

```diff
@@ -398,43 +334,63 @@ func (b *Builder) MachineUIDs() map[string]string {
 // BuildCredentials fetches credential details for current b.Stack from MongoDB.
 //
 // When nil error is returned, the b.Koding and  b.Credentials fields are non-nil.
+//
+// TODO(rjeczalik): Replace with *credential.Client
 func (b *Builder) BuildCredentials(method, username, groupname string, identifiers []string) error {
 	// fetch jaccount from username
 	account, err := modelhelper.GetAccount(username)
 	if err != nil {
-		return ResError(err, "jAccount")
-	}
-
-	// fetch jGroup from group slug name
-	group, err := modelhelper.GetGroup(groupname)
-	if err != nil {
-		return ResError(err, "jGroup")
+		return models.ResError(err, "jAccount")
 	}
 
 	// fetch jUser from username
 	user, err := modelhelper.GetUser(username)
 	if err != nil {
-		return ResError(err, "jUser")
+		return models.ResError(err, "jUser")
 	}
 
-	// validate if username belongs to groupnam
-	selector := modelhelper.Selector{
-		"targetId": account.Id,
-		"sourceId": group.Id,
-		"as": bson.M{
-			"$in": []string{"member"},
-		},
+	kodingMeta := &KodingMeta{
+		Email:     user.Email,
+		Username:  user.Name,
+		Nickname:  account.Profile.Nickname,
+		Firstname: account.Profile.FirstName,
+		Lastname:  account.Profile.LastName,
+		Hash:      account.Profile.Hash,
 	}
 
-	count, err := modelhelper.RelationshipCount(selector)
-	if err != nil || count == 0 {
-		return fmt.Errorf("username '%s' does not belong to group '%s'", username, groupname)
+	groupIDs := []bson.ObjectId{account.Id}
+
+	if groupname != "" {
+		// fetch jGroup from group slug name
```

Comment:
I can also share my long term plans - previous implementation was just a few methods with hundreds of lines:

- https://github.com/koding/koding/commit/6c4720f95a7c3938e0fe3e3d427ca07ad50de6ad#diff-d3b87760e7a4403c45e7dc6e12e714c9L460
- https://github.com/koding/koding/commit/6c4720f95a7c3938e0fe3e3d427ca07ad50de6ad#diff-d3b87760e7a4403c45e7dc6e12e714c9L498
- etc.

I've introduced a Builder type just a temporary abstraction [in this PR](https://github.com/koding/koding/commit/6c4720f95a7c3938e0fe3e3d427ca07ad50de6ad). The purpose of the Builder was to create a base abstraction for validating user, validating credentials, building models, expanding user templates and updating stack models. Basically a facade, that used previous implementation (the global funcs), but stored the product of the methods within itself, so was slightly possible to mock it. The Builder was a first step to fix previous abstraction leak (or lack of abstraction). Right now I'm in an ongoing effort to refactor Builder into separate packages that are responsible for each step of the stack building process: credential, stack and terraform. Once it's done, the Builder will be gone.

The aim is to have functional tests for kloud / stack / template, in addition to integration ones with `kd`.